### PR TITLE
XDLOPS single complex gemm CGEMM

### PR DESCRIPTION
**Pull request by rosenrodt**
- Implemented CGEMM with XDLOPS
- Added respective test config yamls. Local test of `pre_checkin` takes about 1h10m

```md
$ tox -e py36 -- Tensile/Tests -m 'pre_checkin' --durations=0
...
============================== slowest durations ===============================
677.29s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/mfma/hpa_hgemm_asm.yaml]
**442.58s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/mfma/cgemm_asm.yaml]**
353.76s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/mfma/sgemm_asm.yaml]
293.48s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/mfma/hpa_bfloat16_gemm_asm.yaml]
189.69s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/igemm_hpa_hip_tt.yaml]
158.13s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/hgemm_hpa_iu2_asm_nt.yaml]
141.44s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/mfma/wider_local_read.yaml]
134.03s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/igemm_hpa_hip_tn.yaml]
117.92s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/igemm_hpa_hip_nt.yaml]
107.53s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/bfloat16/bfloat16_hpa_source_tt.yaml]
82.48s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/bfloat16/bfloat16_hpa_source_tn.yaml]
**80.22s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/mfma/cgemm_asm_conjugate.yaml]**
74.66s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/bfloat16/bfloat16_hpa_source_nt.yaml]
73.89s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/dgemm_asm.yaml]
43.57s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/source/test_hgemm_defaults.yaml]
43.54s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/source/test_dgemm_defaults.yaml]
42.63s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/source/test_sgemm_defaults.yaml]
41.86s call     Tensile/Tests/yaml_only/test_config.py::test_config[Tensile/Tests/pre_checkin/mfma/1LDSB.yaml]
...
================== 80 passed, 1 skipped in 4430.80s (1:13:50) ==================
___________________________________ summary ____________________________________
  py36: commands succeeded
  congratulations :)
```

### CODE REVIEW

File: Tensile/KernelWriterAssembly.py [hunk 1]
```diff
@@ -5586,16 +5595,46 @@ def mfmaIter(self, kernel, m, innerUnroll, tail=False):
...
+            ccB = kernel["ProblemType"]["ComplexConjugateB"]
+            ccVgprs = [None]*3 # three terms that can be negated: [real1, imag0, imag1]
+            ccInsts = [None]*3
+            accImOffset = self.AccVgprImagNumOffset(kernel)
+            ar = vgpr("ValuA_X%u_I%u+%u+%u+%u"   % (vgprBuffer_new, iui_new, a_new, vgprBuffer_new_offset, iui_new_offset), 1)
+            ai = vgpr("ValuA_X%u_I%u+%u+%u+%u+1" % (vgprBuffer_new, iui_new, a_new, vgprBuffer_new_offset, iui_new_offset), 1)
+            br = vgpr("ValuB_X%u_I%u+%u+%u+%u"   % (vgprBuffer_new, iui_new, b_new, vgprBuffer_new_offset, iui_new_offset), 1)
+            bi = vgpr("ValuB_X%u_I%u+%u+%u+%u+1" % (vgprBuffer_new, iui_new, b_new, vgprBuffer_new_offset, iui_new_offset), 1)
+            v_mfma = "v_mfma_f32_%ux%ux%u%s "%(kernel["MatrixInstM"], kernel["MatrixInstN"], kernel["MatrixInstK"], "f32")
+            if (ccA and ccB) or (not ccA and not ccB):
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:5611]: if ccA == ccB might be simpler?

File: Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -10867,6 +10921,11 @@ def globalWriteBatch(self, kernel, ss, batchIdx, applyAlpha, beta, edge, atomic,
         vc1 = element[2]
         vc0 = element[3]
         sumIdx = ss.elementSumIdx[elementIdx]
+        if self.serializedStore:
+          # ensure index aligned by number of registers per element
+          assert(sumIdx%self.ss.cfg.numVgprPerValuC==0)
+          # convert from index in terms vgprs (serializedStore==True) to index in terms of elements (classic)
+          sumIdx = sumIdx//self.ss.cfg.numVgprPerValuC
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:10928]: Looks the same as the above block, can they be combined?

File: Tensile/KernelWriterAssembly.py [hunk 3]
```diff
@@ -10901,7 +10960,13 @@ def globalWriteBatch(self, kernel, ss, batchIdx, applyAlpha, beta, edge, atomic,
...
           for vi in range(0, gwvw):
             dataV = ss.elementData[elementIdx] + int(vi*ss.cfg.numVgprsPerDataPerVI)
-            sumIdxV = ss.elementSumIdx[elementIdx] + vi
+            sumIdx = ss.elementSumIdx[elementIdx]
+            if self.serializedStore:
+              # ensure index aligned by number of registers per element
+              assert(sumIdx%self.ss.cfg.numVgprPerValuC==0)
+              # convert from index in terms vgprs (serializedStore==True) to index in terms of elements (classic)
+              sumIdx = sumIdx//self.ss.cfg.numVgprPerValuC
+            sumIdxV = sumIdx + vi
```
Comment (AlexBrownAMD) [inline Tensile/KernelWriterAssembly.py:10969]: Repeat

File: Tensile/Common.py [hunk 4]
```diff
@@ -245,7 +245,8 @@
 validMFMA["S"] = [[32,32,1,2], [32,32,2,1], [16,16,1,4], [16,16,4,1], [4,4,1,16]]
 validMFMA["B"] = [[32,32,2,2], [32,32,4,1], [16,16,2,4], [16,16,8,1], [4,4,2,16]]
 validMFMA["4xi8"] = [[32,32,4,2], [32,32,8,1], [16,16,4,4], [16,16,16,1], [4,4,4,16]]
-validMatrixInstructions = [[], [-1]] + validMFMA["H"] + validMFMA["S"] + validMFMA["B"] + validMFMA["4xi8"]
+validMFMA["C"] = validMFMA["S"]
+validMatrixInstructions = [[], [-1]] + validMFMA["H"] + validMFMA["S"] + validMFMA["B"] + validMFMA["4xi8"] + validMFMA["C"]
```
Comment (AlexBrownAMD) [inline Tensile/Common.py:249]: Those matrix instructions are already in the list

File: Tensile/KernelWriterAssembly.py (hunk 1 above)
Comment (rosenrodt) [reply Tensile/KernelWriterAssembly.py]: good idea

File: Tensile/KernelWriterAssembly.py (hunk 2 above)
Comment (rosenrodt) [reply Tensile/KernelWriterAssembly.py]: Thanks for the feedback. The unit conversion (vgpr index to element index) is now done somewhere else so they are all gone now.

**Pull request merged**